File: src/AskLucy.Application/Workflows/Validation/WorkflowDraftDefinition.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using AskLucy.Domain.Workflows;

namespace AskLucy.Application.Workflows.Validation;

/// <summary>Parsed shape of a node inside <see cref="Workflow.DraftDefinitionJson"/> — the Designer's canvas document (research.md Decision 19). Converts 1:1 to a <see cref="WorkflowNodeSpec"/>.</summary>
public sealed record WorkflowDraftNode(
    string NodeKey,
    WorkflowNodeType NodeType,
    string Name,
    string? Description,
    string InputSchemaJson,
    string OutputSchemaJson,
    string ConfigurationJson,
    string RequiredPermissionsJson,
    int? TimeoutSeconds,
    string? RetryPolicyJson,
    WorkflowNodeApprovalPolicy ApprovalPolicy,
    string? IdempotencyKeyExpression,
    string? CompensatingNodeKey,
    double CanvasX,
    double CanvasY)
{
    public WorkflowNodeSpec ToSpec() => new(
        NodeKey, NodeType, Name, Description, InputSchemaJson, OutputSchemaJson, ConfigurationJson,
        RequiredPermissionsJson, TimeoutSeconds, RetryPolicyJson, ApprovalPolicy, IdempotencyKeyExpression,
        CompensatingNodeKey, CanvasX, CanvasY);
}

public sealed record WorkflowDraftConnection(string SourceNodeKey, string TargetNodeKey, string? BranchLabel, string? TypeContract)
{
    public WorkflowConnectionSpec ToSpec() => new(SourceNodeKey, TargetNodeKey, BranchLabel, TypeContract);
}

public sealed record WorkflowDraftVariable(string Name, WorkflowVariableKind Kind, WorkflowVariableType ValueType, string? DefaultValueJson, bool IsRequired)
{
    public WorkflowVariableSpec ToSpec() => new(Name, Kind, ValueType, DefaultValueJson, IsRequired);
}

/// <summary>
/// The parsed shape of <see cref="Workflow.DraftDefinitionJson"/> — shared by <c>ValidateWorkflowCommand</c>
/// and <c>PublishWorkflowVersionCommand</c> so both run against the exact same parsing logic
/// (research.md Decision 19). A malformed draft (invalid JSON) is reported as a single validation
/// issue by <see cref="TryParse"/>'s caller, never an unhandled exception surfaced to the user
/// (constitution §2.VIII).
/// </summary>
public sealed record WorkflowDraftDefinition(
    string InputsSchemaJson,
    string OutputsSchemaJson,
    string ErrorPolicyJson,
    string ExecutionPolicyJson,
    string SecurityPolicyJson,
    IReadOnlyList<WorkflowDraftNode> Nodes,
    IReadOnlyList<WorkflowDraftConnection> Connections,
    IReadOnlyList<WorkflowDraftVariable> Variables)
{
    private static readonly JsonSerializerOptions SerializerOptions = new(JsonSerializerDefaults.Web)
    {
        Converters = { new JsonStringEnumConverter() },
    };

    private sealed record Envelope(
        string? InputsSchemaJson, string? OutputsSchemaJson, string? ErrorPolicyJson, string? ExecutionPolicyJson, string? SecurityPolicyJson,
        List<WorkflowDraftNode>? Nodes, List<WorkflowDraftConnection>? Connections, List<WorkflowDraftVariable>? Variables);

    public static bool TryParse(string draftDefinitionJson, out WorkflowDraftDefinition? definition, out string? parseError)
    {
        try
        {
            var envelope = JsonSerializer.Deserialize<Envelope>(draftDefinitionJson, SerializerOptions);
            if (envelope is null)
            {
                definition = null;
                parseError = "The draft definition is empty.";
                return false;
            }

            definition = new WorkflowDraftDefinition(
                envelope.InputsSchemaJson ?? "{}",
                envelope.OutputsSchemaJson ?? "{}",
                envelope.ErrorPolicyJson ?? "{}",
                envelope.ExecutionPolicyJson ?? "{}",
                envelope.SecurityPolicyJson ?? "{}",
                envelope.Nodes ?? [],
                envelope.Connections ?? [],
                envelope.Variables ?? []);
            parseError = null;
            return true;
        }
        catch (JsonException ex)
        {
            definition = null;
            parseError = $"The draft definition is not valid JSON: {ex.Message}";
            return false;
        }
    }
}

File: src/AskLucy.Web/ClientApp/src/features/workflows/workflowDraftDefinition.ts
import type { WorkflowNodeApprovalPolicy, WorkflowNodeType, WorkflowVariableKind, WorkflowVariableType } from './api/workflowsApi'

/**
 * The exact JSON shape `Workflow.DraftDefinitionJson` round-trips through
 * (`WorkflowDraftDefinition`/`WorkflowDraftNode`/`WorkflowDraftConnection`/`WorkflowDraftVariable`
 * in `src/AskLucy.Application/Workflows/Validation/WorkflowDraftDefinition.cs`) — field names and
 * casing must match verbatim (`System.Text.Json`'s Web defaults camelCase the C# record's
 * PascalCase property names).
 */
export interface WorkflowDraftNodeDefinition {
  nodeKey: string
  nodeType: WorkflowNodeType
  name: string
  description: string | null
  inputSchemaJson: string
  outputSchemaJson: string
  configurationJson: string
  requiredPermissionsJson: string
  timeoutSeconds: number | null
  retryPolicyJson: string | null
  approvalPolicy: WorkflowNodeApprovalPolicy
  idempotencyKeyExpression: string | null
  compensatingNodeKey: string | null
  canvasX: number
  canvasY: number
}

export interface WorkflowDraftConnectionDefinition {
  sourceNodeKey: string
  targetNodeKey: string
  branchLabel: string | null
  typeContract: string | null
}

export interface WorkflowDraftVariableDefinition {
  name: string
  kind: WorkflowVariableKind
  valueType: WorkflowVariableType
  defaultValueJson: string | null
  isRequired: boolean
}

export interface WorkflowDraftDefinition {
  inputsSchemaJson: string
  outputsSchemaJson: string
  errorPolicyJson: string
  executionPolicyJson: string
  securityPolicyJson: string
  nodes: WorkflowDraftNodeDefinition[]
  connections: WorkflowDraftConnectionDefinition[]
  variables: WorkflowDraftVariableDefinition[]
}

export const EMPTY_DRAFT_DEFINITION: WorkflowDraftDefinition = {
  inputsSchemaJson: '{}',
  outputsSchemaJson: '{}',
  errorPolicyJson: '{"strategy":"Stop"}',
  executionPolicyJson: '{}',
  securityPolicyJson: '{}',
  nodes: [],
  connections: [],
  variables: [],
}

/** Tolerant of an empty/unset draft (a brand-new workflow) and of malformed JSON left over from a failed prior save — both fall back to {@link EMPTY_DRAFT_DEFINITION} rather than crashing the Designer. */
export function parseDraftDefinition(draftDefinitionJson: string): WorkflowDraftDefinition {
  if (!draftDefinitionJson || draftDefinitionJson.trim().length === 0) {
    return EMPTY_DRAFT_DEFINITION
  }

  try {
    const parsed = JSON.parse(draftDefinitionJson) as Partial<WorkflowDraftDefinition>
    return {
      inputsSchemaJson: parsed.inputsSchemaJson ?? '{}',
      outputsSchemaJson: parsed.outputsSchemaJson ?? '{}',
      errorPolicyJson: parsed.errorPolicyJson ?? '{"strategy":"Stop"}',
      executionPolicyJson: parsed.executionPolicyJson ?? '{}',
      securityPolicyJson: parsed.securityPolicyJson ?? '{}',
      nodes: parsed.nodes ?? [],
      connections: parsed.connections ?? [],
      variables: parsed.variables ?? [],
    }
  } catch {
    return EMPTY_DRAFT_DEFINITION
  }
}
